src/Unit/AbstractUnit.php

```diff
@@ -42,11 +42,18 @@ public function isMultiInstance(): bool
 
     /**
      * @inheritDoc
+     * @todo: Everything in here should happen inside the constructor and stored
+     *      afterwards.
      */
     public function getInstanceName(): ?string
     {
-        $parts = explode('@', $this->name);
-        return $parts[1] ?? null;
+        $instanceName = explode('@', $this->name, 2)[1] ?? null;
+
+        if (is_string($instanceName) && strpos($instanceName, '.') !== false) {
```

Comment:
Sorry. The test for this one went into the other branch. I have copied it from there. (`AbstractUnitTest::itDetectsMultiInstanceInstanceNamesCorrectly()`)

The problem with your implementation was, that it did not consider service names like 'foo@1.service' etc., which are not prohibited names as for the current implementation.
So the instance name in those cases would have been '1.service', which is obviously not what we want.